### PR TITLE
Qual : collection of methods for common object : Add setter for field value

### PR DESCRIPTION
**Pull request by thersane-john**
# New methods for common object

**setFieldValue**
Set a value for a dynamic field on the current object.
This method performs multiple safety checks before assigning the value:
- verifies that the field is defined
- checks if the user is allowed to edit the field
- optionally validates the value if validation is required

**getRightsForUser**
Returns the rights used for this class for a specific user

**isFieldBlockedByObjectState**
Check if the field is blocked by the current object state.
In non-draft states, fields are locked unless "alwayseditable" is set.

**isFieldEditAllowed**
Check whether a given field is allowed to be edited on the current object.
This method combines:
 - user rights check
 - field configuration constraints
 - object state constraints (status/statut)

 Note: This method is defined in CommonObject, so it must remain generic.
 Some child classes use `status`, others use `statut`, hence both are checked.

**isFieldMarkedNotEditable**
Check if the field is explicitly marked as not editable.
 
**isFieldDefined**
 Check if the field is defined in the object's field metadata.

**isFieldDisabled**
Check if the field is explicitly marked as disabled.

**isFieldValidationRequired**
Check if the field requires validation.

**isFieldEnabled**
Check if the field is enabled.
A field is considered enabled by default if the "enabled" flag is not set.

This pull request is an code extract of #38196

### CODE REVIEW

**Comment by mdeweerd** [comment]
Nice to see some evolution in this area.
May I suggest to use the magic functions?  You can see how https://github.com/yiisoft/yii2/blob/master/framework/base/BaseObject.php uses these.

I did suggest updates in the past (#29145) to refactor the code to use the magic in [DolDeprecationHandler](https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/core/class/doldeprecationhandler.class.php) so that statut automatically resolves to status and even reports on this deprecated usage in the log.
And of course a test case : [DolDeprecationHandlerTest](https://github.com/Dolibarr/dolibarr/blob/develop/test/phpunit/DolDeprecationHandlerTest.php)

**Comment by thersane-john** [comment]
> Nice to see some evolution in this area. May I suggest to use the magic functions? You can see how https://github.com/yiisoft/yii2/blob/master/framework/base/BaseObject.php uses these.

Not yet for `__set()` because I think Dolibarr is not ready for this level of magic handling. I am talking about fields, not just class properties. I prefer to proceed step by step to avoid unexpected side effects. 
However, it could make sense for `__isset()` and `__unset()`. This could be done in a future pull request.

That said, `DolDeprecationHandler` is currently focused on properties, while my changes are targeting fields. So I need to evaluate the impact more carefully and run proper tests before going further in that direction.

---

> I did suggest updates in the past (#29145) to refactor the code to use the magic in [DolDeprecationHandler](https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/core/class/doldeprecationhandler.class.php) so that statut automatically resolves to status and even reports on this deprecated usage in the log.

Thanks for the reference, that is a good point.
I have updated the implementation accordingly.

**Comment by thersane-john** [comment]
@eldy ping

**Comment by mdeweerd** [comment]
> > I did suggest updates in the past (#29145) to refactor the code to use the magic in [DolDeprecationHandler](https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/core/class/doldeprecationhandler.class.php) so that statut automatically resolves to status and even reports on this deprecated usage in the log.
> 
> Thanks for the reference, that is a good point. I have updated the implementation accordingly.

I do not think that there is a need to read the deprecatedProperties as the deprecation handler handles it - just use the new name - that's what the deprecation handler is for.  If the setter is using the old name, then the "new" property will be updated.

This is also something for which a phpunit test can be easily added using AI, so I recommend it.

**Comment by thersane-john** [comment]
@mdeweerd Digging into your suggestion actually surfaced a real bug, so this was worth the extra look.

Regarding relying purely on `DolDeprecationHandler`: it doesn't apply here for `statut`/`status` (or `alreadypaid`/`totalpaid`). Both are still declared as real `public` properties on `CommonObject` (commonobject.class.php#L465-475), and PHP only invokes `__get`/`__set` magic for properties that are undefined or inaccessible, never for a real declared property, regardless of visibility. So for these specific pairs, `DolDeprecationHandler` is effectively inert. It only kicks in for the pairs that were fully removed as real properties (`commandeFournisseur`, `expedition`, `livraison`, `projet`), which are already handled correctly today without any manual code. Manual sync is still required for `statut`/`status` and `alreadypaid`/`totalpaid` until those are actually removed from the class definition, which is a separate, much bigger BC change I'd rather not bundle into this PR.

That said, while re-checking that logic I found an actual bug in the current code: `foreach ($deprecatedProperties as $deprecatedProperty)` iterates over the map's values without ever checking them against `$fieldKey`. In practice this means every single call to `setFieldValue()`, for any field, overwrites `status`, `fk_project`, `totalpaid` and `origin_object` with the value of the field just set (all of these are real properties on `CommonObject`, so the `property_exists()` guard always passes). I've fixed the loop to only sync the pair that actually matches `$fieldKey`, in whichever direction applies.

I've also added PHPUnit tests as you suggested:
- a regression test proving `setFieldValue()` no longer corrupts unrelated properties
- a test proving `statut`/`status` stay in sync in both directions

Pushed both fixes. Let me know if you still think the deprecation-handler-only approach should apply somewhere I'm missing.

**Comment by mdeweerd** [comment]
Declaring properties in "deprecatedProperties" and still having the old property name in place is incorrect.

Your code should not rely on deprecatedProperties, that's the whole point of the DolDeprecationHandler - the magic happens behind the scenes.

There are 2 correct fixes to not rely on deprecatedProperties in your code:
1. Remove (comment) the mappings in deprecatedProperties that still have their older property declared;
2. Remove (comment) the old property declaration.  Really!  That's the way the DolDeprecationHandler can do it's work.  The issue: "human reluctance" - it's never been the "right time", and of course the final word goes to @eldy.

However, as long as DolDeprecationHandler is not correctly used, you kind of need to do it.  I am thinking of adding a test function to DolDeprecationHandler though: 'verifySetup()' which checks in ci that the old property does not exist.

I'll look a bit further in your PR and your specific review question.

File: htdocs/core/class/commonobject.class.php [hunk 1]
```diff
@@ -10177,7 +10177,7 @@ private function isFieldValidationRequired($field)
 	 */
 	private function isFieldEnabled($field)
 	{
-		return !isset($this->fields[$field]['enabled']) || $this->fields[$field]['enabled'];
+		return !isset($this->fields[$field]['enabled']) || (int) dol_eval((string) $this->fields[$field]['enabled']);
```
[the remark this answers is not in the document]
Comment (mdeweerd) [reply htdocs/core/class/commonobject.class.php]: These both return 1:
```bash
php -r 'print (!isset($var["enabled"])?1:0);'
php -r 'print (empty($var["enabled"])?1:0);'
```

(This also points at the typo for 'isempty' and the extra '!' before 'isempty' which actually inverts the logic of 'isset').

A check of the code using `git grep "'enabled'" :*php | grep eval` reveials that most of the time `dol_eval((string) $val['enabled'])` is casted to a bool, and we need a bool, so that should be faster overall.


The compact form merits thought as most of the time `$this->fields[$field]['enabled']` is set (IMHO it avoids an extra lookup of the deep array value).
```php
return (bool) dol_eval((string) ($this->fields[$field]['enabled'] ?? '1'));
```

The "Dolibarr" way would be:
```php
return !isset($this->fields[$field]['enabled']) || (bool) dol_eval((string) $this->fields[$field]['enabled']);
```

**Comment by mdeweerd** [comment]
@thersane-john Review made by AI, I did not check that all statements are true, but good food for thought:

## Review Summary

This PR adds a collection of methods for CommonObject to support field-level validation and editing, including setFieldValue(), isFieldEditAllowed(), and related helper methods. The implementation is generally well-structured with good test coverage.

## Issues Found

### Blocking Issues

[B] isFieldEnabled() returns int instead of bool (commonobject.class.php:10289)
The method casts dol_eval() result to (int), returning 0 or 1 instead of false or true. This violates the PHPDoc @return bool. The cast also has a subtle bug: (int) "true" evaluates to 0 (falsy).

Fix: Change (int) to (bool) or remove cast entirely.

[B] isFieldBlockedByObjectState() logic error (commonobject.class.php:10313)
The condition ($this->status != 0 && $this->statut != 0) is incorrect. Fields should be blocked if EITHER status OR statut is non-zero (not in draft), not only when BOTH are non-zero.

Fix: Change && to ||

[B] Missing translation key (errors.lang)
The code references $langs->trans(FieldNotFoundInObject) but this key is not added to the errors.lang file.

Fix: Add FieldNotFoundInObject = Field not found in object

### Suggestions

[S] Improve isFieldEnabled() implementation (commonobject.class.php:10289)
Consider simplifying to just dol_eval() without casting.

### Nits

[N] TODO comment in getRightsForUser() (commonobject.class.php:10209)
Consider addressing this technical debt.

[N] TODO comment in setValueFrom() (commonobject.class.php:2524)
This should be tracked as a separate issue.

### Praise

[P] Excellent test coverage - comprehensive PHPUnit tests
[P] Good documentation - well-documented with PHPDoc
[P] Thoughtful design - recursion guard is smart

---
Decision: Request changes - Blocking issues need to be addressed before merge.

**Comment by thersane-john** [comment]
Thanks for relaying this, it's a useful checklist even with the disclaimer.

Went through each point against the actual code:

**Confirmed and fixed:**

- `isFieldBlockedByObjectState()`: the `&&` was indeed wrong, changed to `||`. Verified concretely: classes generated by ModuleBuilder (Asset, Workstation, BookCal, and about 35 others) only ever populate `status` and never touch the deprecated `statut`, which stays at its default `null`. Since `null != 0` is `false` in PHP, the `&&` made the whole state check always evaluate to `false` for every one of these classes, so editing was never blocked regardless of the actual object status. `||` fixes this and stays correct for classes that keep both properties in sync (ex: Commande, which always sets `statut` and `status` together). 

**Fixed, but with context:**

- `isFieldEnabled()` returning int instead of bool: the `(int) dol_eval(...)` pattern itself is not something introduced by this PR, it's the exact same pattern already used 5 times in this file (`commonobject.class.php:1094, 5391, 6872, 7167, 9798`) and in `extrafields.class.php` for the same `enabled` flag, so I kept it for consistency rather than removing the cast. Wrapped the final result in `(bool)` so the return value strictly matches the `@return bool` PHPDoc. Added `testIsFieldEnabledReturnsStrictBool`.

**Not changed:**

- TODO in `setValueFrom()` (line 2524): that one is mine, added deliberately. I spotted an unrelated code-location inconsistency there (a product-specific special case living inside a generic method) that deserves its own issue and refactor rather than being folded into this PR. Agree it should stay tracked separately, which is exactly what the TODO is for.
- TODO in `hasUserWritePermissionOnField()` (about `getRightsForUser()`'s mixed `null|int|stdClass` return): agree it's not clean, but it's flagged as a nit even by the review itself, leaving it as tracked debt for now to keep this PR focused.

File: htdocs/core/class/commonobject.class.php [hunk 2]
```diff
@@ -10338,7 +10338,7 @@ private function isFieldValidationRequired($field)
 	 */
 	private function isFieldEnabled($field)
 	{
-		return !isset($this->fields[$field]['enabled']) || (int) dol_eval((string) $this->fields[$field]['enabled']);
+		return !isset($this->fields[$field]['enabled']) || (bool) (int) dol_eval((string) $this->fields[$field]['enabled']);
```
Comment (mdeweerd) [inline htdocs/core/class/commonobject.class.php]: IMHO, just `(bool)`, not `(bool) (int)`

**Comment by mdeweerd** [comment]
@thersane-john
Still WIP.  

AI has enhanced the tests for this PR in my fork. The enhanced tests include:

## Enhanced Tests Added

### 1. Deprecated Property Mappings (Data Provider)
- Tests ALL mappings from deprecatedProperties():
  - statut ↔ status
  - alreadypaid ↔ totalpaid
  - projet ↔ project
  - cond_reglement ↔ depr_cond_reglement

### 2. Field Configuration Tests (Data Provider)
- Tests various field configurations:
  - Enabled fields (default and explicit)
  - Disabled fields
  - Non-editable fields
  - Undefined fields

### 3. Facture Class Integration Tests (Data Provider)
- Tests with actual Facture class (extends CommonObject)
- Tests regular Facture fields: ref, type, total_ht, total_ttc, datef, fk_soc, entity
- Tests deprecated property: statut

### 4. Type-Specific Tests
- Tests onFieldValueChanged hook with various data types:
  - string, integer, float, array, null, empty_string, zero, bool

### 5. State Blocking Tests
- Tests isFieldBlockedByObjectState with various status/statut combinations
- Tests alwayseditable flag behavior

### 6. Expression-Based Enabled Tests
- Tests isFieldEnabled with expression-based enabled flags
- Tests with: "1", "0", "isModEnabled('multicurrency')"

All tests use PHPUnit data providers for comprehensive parameterized testing.

**Location:** mdeweerd/dolibarr branch: test-enhanced-38557
**Commit:** d1d62a86fd30e38fe8da7f2c125250e6b0993d00
**Files changed:** 1 file, +497 lines added to test/phpunit/CommonObjectTest.php

**Comment by mdeweerd** [comment]
@thersane-john Test validation complete!

## Test Results

### Before Enhancement (commit 2a2449c20c6)
- **Tests:** 8
- **Assertions:** 25  
- **Failures:** 0

### After Enhancement (branch: mdeweerd/test-enhanced-38557)
- **Tests:** 30 (+22 new tests)
- **Assertions:** 113 (+88 new assertions)
- **Failures:** 0

## New Tests Added

### 1. Deprecated Property Mappings (6 tests)
Tests all mappings from deprecatedProperties() for properties declared on CommonObject:
- statut ↔ status
- alreadypaid ↔ totalpaid
- projet ↔ project

### 2. Field Configuration Variations (5 tests)
Tests isFieldEditAllowed with various field configurations:
- Field enabled (default)
- Field enabled (explicit '1')
- Field disabled
- Field not editable
- Field not defined

### 3. Facture Class Integration (7 tests)
Tests setFieldValue with actual Facture class (extends CommonObject):
- ref, type, total_ht, total_ttc, datef, fk_soc, entity

### 4. Object State Blocking (1 test)
Tests isFieldBlockedByObjectState with various status/statut combinations and alwayseditable flag.

### 5. Expression-Based Enabled (1 test)
Tests isFieldEnabled with expression-based enabled flags:
- enabled = '1', '0', 'isModEnabled("multicurrency")'

### 6. Type-Specific Tests (1 test)
Tests onFieldValueChanged hook with 8 different data types:
- string, integer, float, array, null, empty_string, zero, bool

## Key Improvements

- **Data Providers:** All major test groups use @dataProvider for comprehensive parameterized testing
- **Real Classes:** Tests use actual Facture class (extends CommonObject) for real-world validation
- **Edge Cases:** Tests cover all deprecated properties, field configurations, and data types
- **Bug Fixes:** Fixed testSetFieldValueDoesNotCorruptUnrelatedProperties to work with new isFieldBlockedByObjectState logic

## Files Modified
- test/phpunit/CommonObjectTest.php (+22 tests, +88 assertions)

**Location:** https://github.com/mdeweerd/dolibarr/tree/test-enhanced-38557
**Commit:** e7ba8955fff

**Comment by mdeweerd** [comment]
I've added multiple tests and with the changes that were made they validate ;-).

I added a PR to your fork.
